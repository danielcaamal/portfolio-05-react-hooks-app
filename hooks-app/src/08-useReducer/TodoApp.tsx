import { TodoList, TodoAdd } from './';
import { useTodo } from '../hooks/useTodo';

export const TodoApp = () => {
    const { todos, todosCount, pendingTodosCount,
        handleDeleteTodo, handleToggleTodo, handleNewTodo } = useTodo();
    
    return (
        <>
            <h1>TodoApp: {todosCount}, <small>pendings: </small>{pendingTodosCount}</h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    <h4>Todos</h4>
                    <hr />
                    <TodoList 
                        todos={todos} 
                        onDeleteTodo={handleDeleteTodo} 
                        onToggleTodo={handleToggleTodo}
                        />
                </div>
                <div className="col-5">
                    <h4>Add Todo</h4>
                    <hr />
                    <TodoAdd 
                        onNewTodo={handleNewTodo} />
                </div>
            </div>
        </>
    ); 
}