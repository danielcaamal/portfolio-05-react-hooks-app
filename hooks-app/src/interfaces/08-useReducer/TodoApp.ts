

export interface TodoInterface {
    id: number;
    description: string;
    done: boolean;
};

export interface TodoListInterface {
    todos: TodoInterface[];
    onDeleteTodo: (id:number) => void;
    onToggleTodo: (id:number) => void;
}

export interface TodoItemInterface {
    todo: TodoInterface;
    onDeleteTodo: (id:number) => void;
    onToggleTodo: (id:number) => void;
}
