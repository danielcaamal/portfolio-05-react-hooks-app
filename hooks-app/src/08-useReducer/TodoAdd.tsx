import { FormEvent } from "react";
import { useForm } from "../hooks/useForm";
import { TodoInterface } from "../interfaces/08-useReducer/TodoApp";

export const TodoAdd = ({ onNewTodo }: { onNewTodo: (todo:TodoInterface) => void }) => {

    const initialTodo = {
        id: 0,
        description: '',
        done: false
    }
    const { formState, onInputChange, onResetForm } = useForm(initialTodo)
    const { description } = formState;

    const onSubmit = (e: FormEvent) => {
        e.preventDefault();

        if (description.trim().length <= 1) return;

        const newTodo = {
            ...initialTodo,
            id: new Date().getTime(),
            description,
        };

        onNewTodo(newTodo);
        onResetForm();
    }

    return (
        <>
            <form
                id="todo-add-form" 
                onSubmit={(e) => onSubmit(e)}>
                <input
                    id="todo-add-input"
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Learn..."
                    autoComplete="off"
                    onChange={onInputChange}
                    value={description}
                />
                <button
                    id="todo-add-button"
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block"
                    >Add
                </button>
            </form>
        </>
    );
}