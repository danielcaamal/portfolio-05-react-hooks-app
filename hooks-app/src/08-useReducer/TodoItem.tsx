import { TodoItemInterface } from "../interfaces/08-useReducer/TodoApp";

export const TodoItem = (
    { 
        todo, 
        onDeleteTodo,
        onToggleTodo 
    } : TodoItemInterface) => {

    return (
        <>
            <span 
                id="todo-item-description"
                className={`align-self-center ${todo.done && 'text-decoration-line-through'}`}
                onDoubleClick={() => onToggleTodo(todo.id)}
                >{todo.description}</span>
            <button
                id="todo-item-delete-button"
                className="btn btn-danger"
                type="button"
                onClick={() => onDeleteTodo(todo.id)}
                >Delete
            </button>
        </>
    );
}