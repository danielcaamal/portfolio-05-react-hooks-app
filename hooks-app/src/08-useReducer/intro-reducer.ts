import { TodoInterface } from "../interfaces/08-useReducer/TodoApp";

const initialState : TodoInterface[] = [
    { id: 1, description: 'Buying a car', done: false },
    { id: 2, description: 'Buying a house', done: false },
]

const todoReducer = (state: TodoInterface[] = initialState, action?: {type?: string, payload: TodoInterface}) => {
    switch (action?.type) {
        case 'add':
            return [...state, action.payload]
        default:
            break;
    }

    return state;
}

let todos = todoReducer(initialState);

const newTodo = {
    id: 3,
    todo: 'Buying a laptop',
    done: false
}

const addTodoAction: any = {
    type: 'add',
    payload: newTodo
}

todos = todoReducer(todos, addTodoAction)