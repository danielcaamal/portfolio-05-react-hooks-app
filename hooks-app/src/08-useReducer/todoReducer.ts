import { TodoInterface } from "../interfaces/08-useReducer/TodoApp";

export const todoReducer = (state: TodoInterface[] = [], action: { type:string, payload: TodoInterface }): TodoInterface[] => {
    switch (action.type) {
        case 'add':
            return [...state, action.payload];
        case 'delete':
            return state.filter(todo => todo.id !== action.payload.id);
        case 'toggle':
            return state.map(
                (todo) => {
                    if (todo.id === action.payload.id) {
                        return {
                            ...todo,
                            done: !todo.done
                        }
                    }
                    return todo;
                }
            )
        default:
            return state;
    }
}