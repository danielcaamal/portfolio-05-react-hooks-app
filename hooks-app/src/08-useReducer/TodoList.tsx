import { TodoListInterface } from '../interfaces/08-useReducer/TodoApp';
import { TodoItem } from './TodoItem';

export const TodoList = (
    {
        todos = [], 
        onDeleteTodo,
        onToggleTodo
    } : TodoListInterface) => {

    return (
        <>
            <ul className="list-group">
                {todos.map((todo, i) => {
                    return (
                        <li 
                            className="list-group-item d-flex justify-content-between" 
                            key={ todo.id }>
                            <TodoItem 
                                todo={todo} 
                                onDeleteTodo={onDeleteTodo}
                                onToggleTodo={onToggleTodo}
                                />
                        </li>
                    )
                })}
            </ul>
        </>
    );
}