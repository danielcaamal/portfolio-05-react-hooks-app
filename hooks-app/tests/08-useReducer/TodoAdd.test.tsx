import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { TodoAdd } from '../../src/08-useReducer/TodoAdd';

describe('TodoAdd', () => {

    const onNewTodo = jest.fn();

    beforeEach(() => jest.clearAllMocks());

    test('should match with snapshot', () => {
        const { container } = render(<TodoAdd onNewTodo={onNewTodo} />);
        expect(container).toMatchSnapshot();
    });

    test('should not call onNewTodo', () => {
        const { container } = render(<TodoAdd onNewTodo={onNewTodo} />);
        const form = container.querySelector('#todo-add-form') as HTMLFormElement;
        fireEvent.submit(form);
        expect(onNewTodo).not.toHaveBeenCalled();
    });

    test('should call onNewTodo', () => {
        const { container } = render(<TodoAdd onNewTodo={onNewTodo} />);
        const input = container.querySelector('#todo-add-input') as HTMLInputElement;
        const form = container.querySelector('#todo-add-form') as HTMLFormElement;
        fireEvent.change(input, { target: { value: 'Learn React' } });
        fireEvent.submit(form);
        expect(onNewTodo).toHaveBeenCalledTimes(1);
    });

    test('should call onNewTodo and reset form', () => {
        const { container } = render(<TodoAdd onNewTodo={onNewTodo} />);
        const input = container.querySelector('#todo-add-input') as HTMLInputElement;
        const form = container.querySelector('#todo-add-form') as HTMLFormElement;
        fireEvent.change(input, { target: { value: 'Learn React' } });
        fireEvent.submit(form);
        expect(onNewTodo).toHaveBeenCalledTimes(1);
        expect(input.value).toBe('');
    });
});