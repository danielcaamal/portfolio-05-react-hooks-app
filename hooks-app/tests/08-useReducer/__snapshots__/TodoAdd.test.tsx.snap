// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoAdd should match with snapshot 1`] = `
<div>
  <form
    id="todo-add-form"
  >
    <input
      autocomplete="off"
      class="form-control"
      id="todo-add-input"
      name="description"
      placeholder="Learn..."
      type="text"
      value=""
    />
    <button
      class="btn btn-outline-primary mt-1 btn-block"
      id="todo-add-button"
      type="submit"
    >
      Add
    </button>
  </form>
</div>
`;
